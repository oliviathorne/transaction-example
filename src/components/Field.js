// @flow
import React from 'react';
import styled from 'styled-components';
import { connect } from 'react-redux';
import { registerField, updateFieldValue } from './../actions/form';

const Label = styled.label`
  display: block;
  width: 100%;
  margin-top: 2.5rem;
  letter-spacing: -0.4px;
`;

const Input = styled.input`
  border: ${(props) => (props.error === true ? '2px #d6d7df' : '2px #d30a19')};
  border-style: none none solid none;
  width: 100%;
  min-height: 2.4rem;
  font-size: 1.1rem;
  caret-color: #4b3cfa;

  &:focus {
    outline: none;
    border-bottom: 2px solid #4b3cfa;
  }

  &[type='number'] {
    padding-left: 18px;
    box-sizing: border-box;
  }
`;

const CurrencyIcon = styled.span`
  position: absolute;
  display: block;
  left: 5px;
  top: 27.5px;
  font-weight: 500;
  font-size: 19px;
  z-index: 1;
`;

const InputContainer = styled.div`
  position: relative;
`;

const ErrorMessage = styled.div`
  color: #d30a19;
  margin-top: 0.5rem;
  font-size: 11pt;
`;

type Props = {
  dispatchRegisterField: (string, string) => void,
  dispatchUpdateFieldValue: (string, string) => void,
  handleSubmit: (?boolean | ?string) => void,
  validate: (string, string, number) => void,
  isSubmiting: boolean,
  showCurrencySymbol: boolean,
  name: string,
  label: string,
  fieldValue: string,
  type: string,
  totalAvailable: number
};

type State = {
  valid: boolean | string
};

class Field extends React.Component<Props, State> {
  constructor(props) {
    super(props);
    this.state = {
      valid: true
    };
  }

  componentWillMount() {
    this.props.dispatchRegisterField(this.props.name, '');
  }

  componentDidUpdate(prevProps) {
    const { isSubmiting, name, fieldValue, validate, totalAvailable } = this.props;

    if (prevProps.isSubmiting !== isSubmiting && isSubmiting) {
      const IS_VALID = validate(name, fieldValue, totalAvailable);
      this.setState({
        valid: IS_VALID
      });

      this.props.handleSubmit(IS_VALID);
    }
  }

  handleChange = (e) => {
    this.props.dispatchUpdateFieldValue(this.props.name, e.target.value);
  };

  handleBlur = () => {
    const { validate, name, fieldValue, totalAvailable } = this.props;
    const VALID = validate(name, fieldValue, totalAvailable);

    this.setState({
      valid: VALID
    });
  };

  render() {
    const { name, label, type, showCurrencySymbol, fieldValue } = this.props;

    return (
      <InputContainer>
        <Label htmlFor={name}>{label}</Label>
        <Input
          name={name}
          type={type}
          value={fieldValue}
          onChange={this.handleChange}
          onBlur={this.handleBlur}
          error={this.state.valid}
        />
        {showCurrencySymbol && <CurrencyIcon>Â£</CurrencyIcon>}
        {this.state.valid !== true && <ErrorMessage>{this.state.valid}</ErrorMessage>}
      </InputContainer>
    );
  }
}

const mapStateToProps = (state, ownProps) => ({
  totalAvailable: state.form.totalAvailable,
  fieldValue: typeof state.form.fields[ownProps.name] !== 'undefined' ? state.form.fields[ownProps.name].value : ''
});

const mapDispatchToProps = (dispatch) => ({
  dispatchRegisterField: (name, initialValue) => dispatch(registerField(name, initialValue)),
  dispatchUpdateFieldValue: (name, value) => dispatch(updateFieldValue(name, value))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Field);
