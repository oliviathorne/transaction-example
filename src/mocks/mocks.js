export const mockState = {
  form: {
    fields: {},
    savedTransactions: [
      {
        name: {
          value: 'Test name'
        },
        email: {
          value: 'email@test.com'
        },
        amount: {
          value: '3000'
        }
      }
    ],
    startingBalance: 18000,
    totalAvailable: 18000,
    totalSent: 0
  }
};
