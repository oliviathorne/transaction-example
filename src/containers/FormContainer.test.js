import { mount } from 'enzyme';
import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import ConnectedFormContainer, { FormContainer } from './FormContainer';
import Field from './../components/Field';
import { mockState } from './../mocks/mocks';

const mockStore = configureStore();
let store;
let wrapper;

beforeEach(() => {
  let validation = jest.fn();

  store = mockStore(mockState);
  wrapper = mount(
    <Provider store={store}>
      <ConnectedFormContainer>
        <Field name="name" label="Name" type="text" validate={validation} />
      </ConnectedFormContainer>
    </Provider>
  );
});

describe('FormContainer component', () => {
  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('renders a child component with isSubmiting and handleSubmit props', () => {
    expect(wrapper.find(Field).length).toBe(1);

    expect(wrapper.find(Field).props().isSubmiting).toBe(false);
    expect(wrapper.find(Field).props().handleSubmit).toBeInstanceOf(Function);
  });

  it('calls the submit method on submit click', () => {
    const unconnectedWrapper = mount(<FormContainer />);
    const form = unconnectedWrapper.find('form');

    form.simulate('submit');
    expect(unconnectedWrapper.state('isSubmiting')).toBe(true);
  });

  it('calls the save transaction method', () => {
    const dispatchSaveTransactionMock = jest.fn();
    const unconnectedWrapper = mount(<FormContainer dispatchSaveTransaction={dispatchSaveTransactionMock} />);

    jest.spyOn(unconnectedWrapper.instance(), 'saveTransaction');
    unconnectedWrapper.instance().saveTransaction();

    expect(unconnectedWrapper.instance().saveTransaction).toHaveBeenCalled();
  });
});
