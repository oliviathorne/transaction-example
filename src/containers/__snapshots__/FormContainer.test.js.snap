// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormContainer component renders correctly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(FormContainer)>
    <FormContainer
      dispatchSaveTransaction={[Function]}
      fields={Object {}}
    >
      <styled.form
        onSubmit={[Function]}
      >
        <StyledComponent
          forwardedClass={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": undefined,
              "componentStyle": ComponentStyle {
                "componentId": "sc-bdVaJa",
                "isStatic": true,
                "lastClassName": "iNmWoD",
                "rules": Array [
                  "
  width: 50%;
  padding: 2rem 10% 4rem 0;
  border-right: 1px solid #edeef7;
  box-sizing: border-box;

  @media (max-width: 768px) {
    width: 100%;
    border-right: 0;
    border-bottom: 1px solid #edeef7;
    padding: 2rem 0 4rem;
  }
",
                ],
              },
              "displayName": "styled.form",
              "render": [Function],
              "styledComponentId": "sc-bdVaJa",
              "target": "form",
              "toString": [Function],
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            }
          }
          forwardedRef={null}
          onSubmit={[Function]}
        >
          <form
            className="sc-bdVaJa iNmWoD"
            onSubmit={[Function]}
          >
            <styled.h1>
              <StyledComponent
                forwardedClass={
                  Object {
                    "$$typeof": Symbol(react.forward_ref),
                    "attrs": undefined,
                    "componentStyle": ComponentStyle {
                      "componentId": "sc-bwzfXH",
                      "isStatic": true,
                      "lastClassName": "hllhvC",
                      "rules": Array [
                        "
  font-family: 'AlverataBlack';
  letter-spacing: -0.6px;
",
                      ],
                    },
                    "displayName": "styled.h1",
                    "render": [Function],
                    "styledComponentId": "sc-bwzfXH",
                    "target": "h1",
                    "toString": [Function],
                    "warnTooManyClasses": [Function],
                    "withComponent": [Function],
                  }
                }
                forwardedRef={null}
              >
                <h1
                  className="sc-bwzfXH hllhvC"
                />
              </StyledComponent>
            </styled.h1>
            <Connect(Field)
              handleSubmit={[Function]}
              isSubmiting={false}
              key=".0"
              label="Name"
              name="name"
              type="text"
              validate={[MockFunction]}
            >
              <Field
                dispatchRegisterField={[Function]}
                dispatchUpdateFieldValue={[Function]}
                fieldValue=""
                handleSubmit={[Function]}
                isSubmiting={false}
                label="Name"
                name="name"
                totalAvailable={18000}
                type="text"
                validate={[MockFunction]}
              >
                <styled.div>
                  <StyledComponent
                    forwardedClass={
                      Object {
                        "$$typeof": Symbol(react.forward_ref),
                        "attrs": undefined,
                        "componentStyle": ComponentStyle {
                          "componentId": "sc-bZQynM",
                          "isStatic": true,
                          "lastClassName": "iMWVsx",
                          "rules": Array [
                            "
  position: relative;
",
                          ],
                        },
                        "displayName": "styled.div",
                        "render": [Function],
                        "styledComponentId": "sc-bZQynM",
                        "target": "div",
                        "toString": [Function],
                        "warnTooManyClasses": [Function],
                        "withComponent": [Function],
                      }
                    }
                    forwardedRef={null}
                  >
                    <div
                      className="sc-bZQynM iMWVsx"
                    >
                      <styled.label
                        htmlFor="name"
                      >
                        <StyledComponent
                          forwardedClass={
                            Object {
                              "$$typeof": Symbol(react.forward_ref),
                              "attrs": undefined,
                              "componentStyle": ComponentStyle {
                                "componentId": "sc-bxivhb",
                                "isStatic": true,
                                "lastClassName": "hpIQsY",
                                "rules": Array [
                                  "
  display: block;
  width: 100%;
  margin-top: 2.5rem;
  letter-spacing: -0.4px;
",
                                ],
                              },
                              "displayName": "styled.label",
                              "render": [Function],
                              "styledComponentId": "sc-bxivhb",
                              "target": "label",
                              "toString": [Function],
                              "warnTooManyClasses": [Function],
                              "withComponent": [Function],
                            }
                          }
                          forwardedRef={null}
                          htmlFor="name"
                        >
                          <label
                            className="sc-bxivhb hpIQsY"
                            htmlFor="name"
                          >
                            Name
                          </label>
                        </StyledComponent>
                      </styled.label>
                      <styled.input
                        error={true}
                        name="name"
                        onBlur={[Function]}
                        onChange={[Function]}
                        type="text"
                        value=""
                      >
                        <StyledComponent
                          error={true}
                          forwardedClass={
                            Object {
                              "$$typeof": Symbol(react.forward_ref),
                              "attrs": undefined,
                              "componentStyle": ComponentStyle {
                                "componentId": "sc-ifAKCX",
                                "isStatic": false,
                                "lastClassName": "czOlHA",
                                "rules": Array [
                                  "
  border: ",
                                  [Function],
                                  ";
  border-style: none none solid none;
  width: 100%;
  min-height: 2.4rem;
  font-size: 1.1rem;
  caret-color: #4b3cfa;

  &:focus {
    outline: none;
    border-bottom: 2px solid #4b3cfa;
  }

  &[type='number'] {
    padding-left: 18px;
    box-sizing: border-box;
  }
",
                                ],
                              },
                              "displayName": "styled.input",
                              "render": [Function],
                              "styledComponentId": "sc-ifAKCX",
                              "target": "input",
                              "toString": [Function],
                              "warnTooManyClasses": [Function],
                              "withComponent": [Function],
                            }
                          }
                          forwardedRef={null}
                          name="name"
                          onBlur={[Function]}
                          onChange={[Function]}
                          type="text"
                          value=""
                        >
                          <input
                            className="sc-ifAKCX czOlHA"
                            name="name"
                            onBlur={[Function]}
                            onChange={[Function]}
                            type="text"
                            value=""
                          />
                        </StyledComponent>
                      </styled.input>
                    </div>
                  </StyledComponent>
                </styled.div>
              </Field>
            </Connect(Field)>
            <styled.input
              type="submit"
            >
              <StyledComponent
                forwardedClass={
                  Object {
                    "$$typeof": Symbol(react.forward_ref),
                    "attrs": undefined,
                    "componentStyle": ComponentStyle {
                      "componentId": "sc-htpNat",
                      "isStatic": true,
                      "lastClassName": "dRgDSk",
                      "rules": Array [
                        "
  width: 85%;
  display: block;
  margin: 8rem auto 0;
  background-color: #4b3cfa;
  border: 0;
  color: #ffffff;
  height: 3rem;
  border-radius: 1.5rem;
  font-size: 16px;
  font-weight: 600;

  &:focus {
    outline: none;
  }
",
                      ],
                    },
                    "displayName": "styled.input",
                    "render": [Function],
                    "styledComponentId": "sc-htpNat",
                    "target": "input",
                    "toString": [Function],
                    "warnTooManyClasses": [Function],
                    "withComponent": [Function],
                  }
                }
                forwardedRef={null}
                type="submit"
              >
                <input
                  className="sc-htpNat dRgDSk"
                  type="submit"
                />
              </StyledComponent>
            </styled.input>
          </form>
        </StyledComponent>
      </styled.form>
    </FormContainer>
  </Connect(FormContainer)>
</Provider>
`;
